Update .travis.yml to test Python 3.5

asmeurer opened the pull request:
Fixes https://github.com/sympy/sympy/issues/9927

I also will fix any test failures that exist for Python 3.5, and try to reduce the number of builds in Travis (this bumps the total from 43 to 48 builds). 

**TODO**
- [x] Check every Travis build to make sure it is doing the right thing
- [x] Double check no tests were removed from Travis
- [x] Fix Sage tests on Travis
- [x] Fix all test failure on Travis
- [x] Test allowed failures
- [x] See if there is some way to test against a newer PyPy (which hopefully won't crash as much)

asmeurer committed:
Update .travis.yml to test Python 3.5

asmeurer committed:
Merge several builds together in Travis

I also reorganized the test_travis.sh script so that it can run multiple
things for different environment variables at once.

asmeurer committed:
Fix yaml in .travis.yml

asmeurer committed:
Skip infinite recursion example doctests in the gotchas

In Python 3.5, recursion errors raise RecursionError instead of
RuntimeError. Instead of using ellipses, which would just make the example
harder to read, we just skip the doctest.

asmeurer committed:
Explicitly disallow lambda with starred args in sympy_tokenize

Previously it was relying on the generated Lambda having invalid syntax, but
the syntax of the generated Lambda is valid in Python 3.5 (but
nonsensical). Instead, we detect and disallow this explicitly in the
tokenizer. This also allows us to give a better error message.

asmeurer committed:
test_travis.sh: don't attempt to create an empty directory twice

asmeurer committed:
Travis: test against the latest versions of NumPy, SciPy, and Matplotlib

asmeurer committed:
Make sure the normal tests are actually run in Travis

asmeurer committed:
Fix syntax in Travis test file

asmeurer committed:
Travis: don't run the full tests in the PyPy doctests build

asmeurer committed:
Print the name of the doctest in verbose mode

asmeurer committed:
Fix printing of skipped tests in verbose mode

Also, print what the doctest depends on for verbose mode for doctests.

asmeurer committed:
Print a more direct verbose message for doctest_depends_on skips

This changes the internal API of _process_dependencies to return True or a
message string instead of returning True or False.

asmeurer commented:
Various warnings printed in the matplotlib doctests of the form 

```
/Users/aaronmeurer/anaconda3/lib/python3.5/site-packages/matplotlib/collections.py:590: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  if self._edgecolors == str('face'):
```

are due to a bug in matplotlib (https://github.com/matplotlib/matplotlib/issues/5209) and I've verified that they go away development version (1.5.0).

isuruf commented:
New docker workers for travis can be used for all tests now as all apt dependencies are now whitelisted by travis except for sage (which can be installed by downloading the tar and unpacking like [here](https://github.com/symengine/symengine.py/blob/19b77b40987882f8538f50d4948bf031e1237030/bin/install_travis.sh#L22))

Two allowed failure tests are not run as they are not duplicated in build matrix (https://github.com/sympy/sympy/pull/9792)

asmeurer commented:
I was wondering why the allowed failures builds weren't running.

asmeurer committed:
Clean up some imports in the pretty printing tests

asmeurer committed:
Fix pretty printing of MatrixSlice

It was calling stringpict.__len__, which referenced a non-existent
self.s. Some basic tests for pretty printing MatrixSlice were added.

asmeurer committed:
Fix theano tests with the latest version of theano

asmeurer committed:
Fix doctests in Python 2.6

'{}'.format('') fails in Python 2.6, so we just use regular formatting
instead.

asmeurer committed:
Don't cache sympy.physics.vector.frame.CoordinateSym

The default Symbol.__new__ is cached, but only based on the name and
assumptions. CoordinateSym.__new__ takes two additional arguments, which were
not being considered as part of the cache. This was causing problems when
creating two ReferenceFrame objects with the same name (since ReferenceFrame
objects are technically mutable, they are hashed by the default
object.__hash__, i.e., by id()). This was manifesting itself in errors in
Python 3.5 due to a bug in its lru_cache implementation
(https://bugs.python.org/issue25295), because the physics rst doctests were
creating multiple ReferenceFrame('R') instances (which themselves create
CoordinateSym instance that reference them).

Technically, there are still a lot of problems here. ReferenceFrame shouldn't
be mutable, or if it is, it shouldn't be used as if it were immutable. It
really should be an immutable subclass of Basic and used in CoordinateSym's
.args (which shouldn't subclass from Symbol). I believe these issues are being
addressed in the new sympy.vector implementation, so I'm not too worried about
it.

More worrisome is that any subclass of Symbol that attempts to add new
properties to its identity (i.e., custom __eq__ and __hash__) will run into
this issue. Such subclasses will need to work around this either by disabling
caching (doing the same thing that I have done here), or by including the
extra properties in the assumptions, which are included as part of the Symbol
hash (but assumptions must be booleans, so this does limit things). Even
better, subclass Expr instead of Symbol, and put your extra properties in
.args if possible.

I don't know if there is a good way to fix this from the Symbol side so that
this situation is avoided automatically. Regardless, better documentation on
how to properly subclass Symbol and how to create custom SymPy objects is
definitely needed.

I seriously doubt this will have any performance impact on real code, since
"correct" code should only be creating one ReferenceFrame for any given name.

asmeurer commented:
@moorepants @srjoglekar246 you may want to take a look at https://github.com/sympy/sympy/commit/47cfa23f48bc5274ecdbd17ef30403478d2c2b57. It took me a while to figure out what was going on, so I don't expect you to understand it all, but it would be nice to at least get a confirmation that that code is indeed being replaced by sympy.vector (which looked like it was written much better from what I saw).

asmeurer committed:
Don't halt the test runner if the test raises an exception

There is already a method for the test to be marked as raising an exception.

This prints an extra E before the test file name for some reason which I can't
figure out, but at least the tests aren't halted any more.

asmeurer committed:
Fix printing of vector objects with other objects

The implementation was creating some kind of fake prettyForm object instead of
just using prettyForm (for no good reason), and it didn't even subclass it,
hence the errors. The BasisDependent printer now just returns a normal
prettyForm like every other printer.

Fixes #10044.

asmeurer commented:
OK, actually that was a lie that things are written well in sympy.vector. They are better (in the sense that all objects are at least subclassing from Basic), but there are a lot of issues. Fixes are incoming, and I would appreciate a review from one or both or you.

asmeurer committed:
Make some error messages more helpful

asmeurer committed:
First pass a making BaseScalar subclass from Expr instead of Symbol

It subclassing from Symbol causes the same issues described in the commit
message of 47cfa23f48bc5274ecdbd17ef30403478d2c2b57.

I have made the arguments of the object actually use .args.

So far there are some test failures because of printing changes, and also some
failures due to sympy.vector.Point not being recreatable from its .args.

asmeurer committed:
Merge branch 'master' into python35

There are still some test failures with the BaseScalar but I will fix them in
later commits.

asmeurer committed:
Fix some vector tests

asmeurer committed:
Make BaseScalar commutative

asmeurer committed:
Fix use of atoms in sympy.vector.express

asmeurer committed:
Build the Sphinx docs using Python 3.5 in Travis

asmeurer commented:
@isuruf are there disadvantages to having the one Sage build use sudo? I would rather install it using apt than from a download as then it will keep up-to-date to the latest version (or at least the latest version in the repos, but that will likely be updated more often than a hard-coded link in our test configuration). 

The sage tests run very fast, and my guess is that taking a little longer to spin up won't make a huge difference.

asmeurer committed:
Remove sudo: true for autowrap and sphinx Travis tests

asmeurer committed:
Fix indentation in .travis.yml

isuruf commented:
Only disadvantage is that using apt-get fails quite regularly. (Downloading the tar and extracting fails very rarely). Since it is an allowed failure test, that doesn't matter though.

Upabjojr commented on sympy/vector/scalar.py:
```diff
@@ -13,11 +13,17 @@ class BaseScalar(Symbol):
+
+        index = _sympify(index)
+        system = _sympify(system)
+        obj = super(BaseScalar, cls).__new__(cls, Symbol(name), index, system, Symbol(pretty_str), Symbol(latex_str))
```
`name` should be contained in `system`, I guess.

Upabjojr replied:
https://github.com/asmeurer/sympy/blob/6ded1d78dd0bfb457cf2036416c35bce9daad8d1/sympy/vector/coordsysrect.py#L17

`variable_names` is passed as argument to `CoordSysCartesian`, but not registered as _args_ there. It gets passed as `name` here. I think it should be stored in `CoordSysCartesian`'s _args_, so that `name` would be available from `system` here.

That's the way @Krastanov did it in _sympy.diffgeom_, it looks neater.

Same argument for vector fields.

asmeurer replied:
I based this on @srjoglekar246's changes from https://github.com/sympy/sympy/pull/10130.

asmeurer replied:
If you have better fixes, pull requests are welcome, either to this branch, or to master (I can merge this branch once they are in). I don't have much knowledge on how this stuff works, so I'm not really in a position to fix design issues, only simple bugs.

srjoglekar246 replied:
That would be a good way to go about it, @Upabjojr . I will open a PR with it in a day or two.

srjoglekar246 replied:
This will work without any issues @asmeurer . @Upabjojr 's way would just make the code look cleaner.

asmeurer replied:
I'm unclear what exactly @Upabjojr is suggesting to do. Can you give an example of what the args for CoordSysCartesian and BaseScalar would look like under your suggestion?

srjoglekar246 replied:
With that suggestion, CoordSysCartesian's args would include the names given to its base vectors and scalars. As a result, we would not need to pass the string forms (l.e. the args Symbol(pretty_str), Symbol(latex_str) in the code above) during the construction of BaseScalar. They would be available from the corresponding system instance, which is already known to BaseScalar.

asmeurer committed:
Merge branch 'master' into python35

asmeurer committed:
Don't use atoms() with no arguments in sympy.vector

Instead, use atoms(BaseScalar), or whatever class we are looking for. This was
breaking because BaseScalar was no longer a subclass of Atom.

asmeurer committed:
Make integrate work with _diff_wrt expressions

They work with diff, so it makes sense that they should also work with
integrate().

asmeurer committed:
Merge branch 'integrate-diff_wrt' into python35

asmeurer committed:
Add _eval_derivative to sympy.vector.BaseScalar

This is needed to make derivatives like x.diff(x) evaluate. See also issue #10150.

asmeurer committed:
Fix a test failure due to output differences

asmeurer committed:
sympy.vector: always trigsimp the matrix in orient_new

The matrix is trigsimped in scalar_map, which causes the creation of
CoordinateSystem objects that appear the same, but compare unequal because
their matrices are in different forms.

See https://groups.google.com/d/msg/sympy/Jxpadu27YME/hR8bkGSCCgAJ for more
information.

asmeurer committed:
Fix some simple vector printing test failures

They were caused by expression re-ordering due to the recent change in the
args structure of BaseScalar.

asmeurer committed:
Add a TODO about calling trigsimp

asmeurer committed:
Replace an instance of _needs_brackets in the LaTeX printer with parenthesize

See #10157. This was needed to fix an issue with the vector printer. The
ScalarBase object needs to be parenthesized, but _needs_brackets doesn't allow
this level of customization (whereas parenthesize does).

asmeurer committed:
See the precedence of ScalarBase

This makes it parenthesize correctly when printed.

asmeurer committed:
Fix vector doctests due to printer reordering

asmeurer committed:
Add BaseScalar.name

Some functions in the core expect objects from free_symbols to have .name.

asmeurer committed:
Clear trailing whitespace in doc/src/modules/vector/examples.rst

asmeurer committed:
Fix some quantum printing tests

This removes some parentheses from the latex form, but the pretty printed form
already didn't have parentheses (and they don't seem to be necessary).

asmeurer committed:
Fix latex printing tests of Jz in the quantum module

This is a slight regression, in that Jz**2 now prints as J_z^{2} instead of
\left(J_z\right)^{2}, which doesn't match the other pretty printers. However,
due to the way precedence works in the printers, there is only one precedence,
and making it print with precedence that parenthesizes it also causes it to be
parenthesized for the string printer, which doesn't need it.  I chose to go
with this behavior because didn't require changing any of the quantum
code. The other way can be achieved by adding

    from sympy.printing.precedence import PRECEDENCE
    precedence = PRECEDENCE[Pow']

to JxOp (and similar classes).

A better fix would be to refactor the printers so that precedence can be set
separately for different printers.

asmeurer committed:
Fix a vector printing test

The redundant parentheses are no longer printed.

asmeurer committed:
Correctly use _diff_wrt instead of is_Symbol in Derivative._eval_subs

Fixes a broken vector doctest. Changed a test that was testing that
Derivative(f(x), x).subs(x, g(x)) returned a Subs and added a new test for an
instance that still returns a Subs.

asmeurer committed:
Test against the latest version of theano in Travis

asmeurer committed:
Fix apt installs for Travis Sphinx build

asmeurer committed:
Travis: Install scipy with verbose mode

Otherwise it can take more than 10 minutes to give output, which causes Travis
to error the job.

asmeurer commented:
Is it common for the slow tests to time out like that?

asmeurer committed:
Try adding more verbosity to the scipy install

asmeurer committed:
Travis: Try installing scipy with apt

asmeurer committed:
Try installing scipy, matplotlib, and numpy in Travis with conda

asmeurer commented:
Nice, installing stuff with conda worked. Now I just need to fix the "RuntimeError: Invalid DISPLAY variable" errors from matplotlib.

asmeurer committed:
Use the Agg matplotlib backend when doctesting

This should fix the "Invalid DISPLAY variable" when testing on Travis CI. See
https://stackoverflow.com/questions/4706451/how-to-save-a-figure-remotely-with-pylab.

asmeurer committed:
Add flush() to MockPrint in the plot tests

asmeurer commented:
A little disconcerting that the test errors are showing up at https://travis-ci.org/sympy/sympy/jobs/103425545 but not at https://travis-ci.org/sympy/sympy/jobs/103425544.

asmeurer committed:
Add encoding to MockPrint in the plotting tests

asmeurer committed:
Fix a plotting test failure (#8577)

There was already code for handling unorderable complex types from SymPy but
it didn't handle the error from Python (3) itself.

asmeurer commented:
The very last test failure (at least I'm hoping it's the last one) is 

```
In [1]: plot_implicit(y - cos(pi / x))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-1-c7cb69022999> in <module>()
----> 1 plot_implicit(y - cos(pi / x))

/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/plotting/plot_implicit.py in plot_implicit(expr, x_var, y_var, **kwargs)
    372     p = Plot(series_argument, **kwargs)
    373     if show:
--> 374         p.show()
    375     return p

/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/plotting/plot.py in show(self)
    183             self._backend.close()
    184         self._backend = self.backend(self)
--> 185         self._backend.show()
    186
    187     def save(self, path):

/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/plotting/plot.py in show(self)
   1016
   1017     def show(self):
-> 1018         self.process_series()
   1019         #TODO after fixing https://github.com/ipython/ipython/issues/1255
   1020         # you can uncomment the next line and remove the pyplot.show() call

/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/plotting/plot.py in process_series(self)
    919                 self.ax.spines['left'].set_smart_bounds(False)
    920                 self.ax.spines['bottom'].set_smart_bounds(False)
--> 921                 points = s.get_raster()
    922                 if len(points) == 2:
    923                     #interval math plotting

/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/plotting/plot_implicit.py in get_raster(self)
     78         yinterval = interval(self.start_y, self.end_y)
     79         try:
---> 80             temp = func(xinterval, yinterval)
     81         except AttributeError:
     82             if self.use_interval_math:

<string> in <lambda>(x0, x1)

TypeError: unsupported operand type(s) for /: 'Pi' and 'interval'

```

This doesn't show up in the Travis logs for some reason, but it does when I test locally. I think perhaps the stdout mocking is screwing things up (maybe this should be considered its own test issue to be fixed). 

It is related to the change of printing `Eq` as `Eq` (instead of using `==`). This breaks experimental_lambdify, because the code how takes https://github.com/sympy/sympy/blob/ae320c4ad88c7aa2c35d35fc3489768cd88d8eb2/sympy/plotting/experimental_lambdify.py#L560 instead of https://github.com/sympy/sympy/blob/ae320c4ad88c7aa2c35d35fc3489768cd88d8eb2/sympy/plotting/experimental_lambdify.py#L558 for `Eq` objects, and the `pi` -> `math.pi` translation only happens in `translate_str`. 

I thought the fix might be to call `translate_str` on `tree[1]`, [here](https://github.com/sympy/sympy/blob/ae320c4ad88c7aa2c35d35fc3489768cd88d8eb2/sympy/plotting/experimental_lambdify.py#L560), but that just causes the plot to take a really long time to come up, and to be very grainy (although otherwise correct).  Maybe the fix is to do the replacement somewhere in `translate_func`. Or maybe `interval` should learn how to do operations with `pi` (that option seems wrong to me given the design of the module). 

Just tossing this out there in case someone else has any thoughts.

asmeurer committed:
Merge branch 'master' into python35

asmeurer committed:
Recursively translate in experimental_lambdify

This fixes a test failure "TypeError: unsupported operand type(s) for /: 'Pi'
and 'interval'" from plot_implicit(y - cos(pi / x)). The failure can be traced
to the changing of the printing of Eq(x, y) from "x == y" to "Eq(x, y)". This
causes experimental_lambdify() to take a different code path, based on a
function call instead of an expression. Translations like sympy.pi -> math.pi
were only happening in the translate_str() function. They are now done in
translate_func(). The interval arithmetic for plot_implicit() only works for
floats, hence the error message.

Guards to disable float_wrap_evalf and complex_wrap_evalf are needed here
because we don't want to wrap internal expressions in a function with float()
or complex() (this breaks plotting of Integral(log((sin(x)**2 + 1)*sqrt(x**2 + 1)),
(x, 0, y)) in the tests, for instance, because the internal symbolic
expressions in x are wrapped in float() or complex(), but x is a symbolic
bound symbol that only becomes numeric when the integral is computed).

asmeurer committed:
Don't translate oo to 1e400 in experimental_lambdify()

This causes a test of plotting an infinite summation to hang, because evalfing
an infinite sum is much faster than evalfing a large finite one.

As far as I can tell, matplotlib handles float('inf') just fine, so I don't
see the need to translate this.

asmeurer committed:
Use Sum instead of summation in plotting tests

According to the comments, this is clearly supposed to test a Sum. But
summation was computing the sum in closed form for at least one instance here
(it was testing plotting harmonic(y) instead of the Sum). This makes the test
a little slower, but I believe it is what was intended to be tested. In the
other instance the sum was not computed in closed form but could be in the
future (it's the zeta function).

asmeurer committed:
Close plots in plot_implicit tests

Otherwise matplotlib gives a warning for too many open plots that messes up
the test that tests for warnings.

asmeurer committed:
Fix the plot_implicit warning test

The warning uses the spelling labelled instead of labeled.

asmeurer committed:
Save the experimental_lambdify string on the object

This makes debugging easier.

asmeurer committed:
Close plot.py tests

This avoids a warning from matplotlib about too many concurrent plots.

asmeurer committed:
Print what is testing in each build in Travis

asmeurer committed:
Fix testing logic in test_travis.sh

asmeurer committed:
Try setting matplotlib.use("Agg") in test_travis.sh

asmeurer committed:
Use subprocess=False for the matplotlib tests

asmeurer commented:
What the heck is going on?

<img width="1067" alt="screenshot 2016-01-27 22 41 46" src="https://cloud.githubusercontent.com/assets/71486/12634425/4ebc30ee-c547-11e5-8c86-30bbc55cb2af.png">

```bash
sympypython35%$git log python35 | head -n 20
commit 53c8505430884a87b503c76fb59bd0853a26739f
Author: Aaron Meurer <asmeurer@gmail.com>
Date:   Wed Jan 27 20:04:31 2016 -0500

    Use subprocess=False for the matplotlib tests

commit 62ec53069f9bc3f28719b48ca73e220ee3694a94
Author: Aaron Meurer <asmeurer@gmail.com>
Date:   Wed Jan 27 15:57:03 2016 -0500

    Try setting matplotlib.use("Agg") in test_travis.sh

commit 9b396ad42b986720e9de82ef584b79c1f6c850aa
Author: Aaron Meurer <asmeurer@gmail.com>
Date:   Wed Jan 27 11:05:22 2016 -0500

    Fix testing logic in test_travis.sh

commit 7243e0030616d6caea1a75a56dce3597c86ef4b8
Author: Aaron Meurer <asmeurer@gmail.com>
sympypython35%$git push github python35
Everything up-to-date
```

asmeurer commented:
I guess GitHub was down and it broke it. Hopefully it will fix itself the next time I make a commit to push.

asmeurer committed:
Add PyPy tests to the Travis allowed failures

They fail randomly because of PyPy bugs and I'm sick of it.

asmeurer committed:
Actually run the Sage tests on Travis

Allowed failures builds have to be duplicated in the test matrix to run

asmeurer commented:
Good it's working now.

asmeurer commented:
I've added PyPy to the allowed failures, because I'm tired of it failing because of random PyPy bugs (one of the builds was already supposed to be there but it wasn't configured properly).

asmeurer committed:
Add missing PyPy SPLIT=4/4 build to Travis

asmeurer commented:
Travis failed because of https://github.com/travis-ci/travis-ci/issues/5492. I'm restarting the job.

asmeurer commented:
I downloaded the latest version of Sage (newer than what is on Travis) and the tests pass for me with it. So I am going to ignore the Sage Travis failures for now.

asmeurer commented:
Actually there is a newer version of Sage in our ppa (https://launchpad.net/~aims/+archive/ubuntu/sagemath/+packages), but Travis is downloading the older one because it is listed as "precise", which is the version of ubuntu it is running. How do I tell apt-get to get the newer version?

asmeurer commented:
Actually, now that I think about it, that probably won't work. Linux binaries are not forwards compatible. So I guess we'll just have to wait until either the ppa updates their precise binaries or Travis updates their Ubuntu. Unless there is another ppa we can use with newer precise Sage binaries.

moorepants commented:
It's probably unlikely that the ppa will make precise binaries since there is now a new LTS (14.04) out. Travis does have the option to switch to 14.04 I believe. I've yet to do that for any of my builds.

asmeurer commented:
I see. We can do it, but it won't run on the container infrastructure (meaning it will be slower to start up). See https://docs.travis-ci.com/user/ci-environment/#Virtualization-environments.

asmeurer commented:
Oh it already runs with `sudo: true`. So let me see if this works.

asmeurer committed:
Run the Sage tests on a newer Ubuntu version in Travis

This will let the ppa use the newest version of Sage, which is what we want to
test against.

See also
https://docs.travis-ci.com/user/ci-environment/#Virtualization-environments.

asmeurer committed:
Add -y to the apt-get install command in the Sage Travis build

asmeurer committed:
Remove Sage from Travis allowed failures

asmeurer committed:
Fix deprecated Sage function behavior

asmeurer committed:
Add note to a sage XFAIL test (it should remain XFAILed, even though the test passes)

asmeurer commented:
This PR is ready to be reviewed and merged.

moorepants commented on .travis.yml:
```diff
@@ -62,89 +86,128 @@ matrix:
+        conda update -q conda;
+        conda info -a;
+        conda create -q -n test-environment python=$TRAVIS_PYTHON_VERSION matplotlib numpy scipy pip;
+        source activate test-environment;
```
Yes!

moorepants commented on sympy/vector/coordsysrect.py:
```diff
@@ -438,6 +438,12 @@ def orient_new(self, name, orienters, location=None,
+            # end up with the same CoordinateSystem that compares differently
+            # due to a differently formatted matrix). However, this is
+            # probably not so good for performance.
+            final_matrix = trigsimp(final_matrix)
```
This is probably really bad for performance. I'd prefer not to have it in there, but we can merge this and figure that out later. I don't think the new vector package is used for any performance critical tasks yet, e.g. we don't use it for the mechanics package.

asmeurer replied:
Yeah, there is a design issue to be worked out here, and I'm not the one to do it. The issue is that internal matrices are being rewritten (using trigsimp, i.e., in scalar_map), but then used for equality comparison. You'll need to think about how you want to do canonicalization and/or equality testing.

moorepants replied:
Ok, maybe we can get some people to work on this this summer.

moorepants commented on .travis.yml:
```diff
@@ -62,89 +86,128 @@ matrix:
+            - python-numpy
+            - cython
       virtualenv:
         system_site_packages: true
```
I think this is irrelevant when you are using conda.

moorepants replied:
We should probably disable Travis's virtualenv stuff if possible.

asmeurer replied:
Only the one set of builds use conda. There is no gmpy for conda, so we can't use it everywhere (and I don't have a good Linux building VM to build one myself).

moorepants commented on .travis.yml:
```diff
@@ -62,89 +86,128 @@ matrix:
   - if [[ "${TEST_AUTOWRAP}" == "true" ]]; then
       pip uninstall -y numpy;
-      sudo apt-get update;
-      sudo apt-get install gfortran gcc python-numpy cython;
```
Where is cython installed for the autowrap tests?

asmeurer replied:
[Here](https://github.com/sympy/sympy/pull/10084/files#diff-354f30a63fb0907d4ad57269548329e3R111). .travis.yml has specific syntax for installing apt packages.

moorepants replied:
ok, cython could also be a conda install.

asmeurer replied:
Yeah we could probably do this build with conda too. 

Once there is a usable gmpy package we should move all the builds over to using conda. For now, I *have* to do the scipy builds with conda because pip installing it is too slow.

moorepants commented on .travis.yml:
```diff
@@ -62,89 +86,128 @@ matrix:
-      pip install "numpy==1.7.1";
-      pip install "scipy==0.12.0";
-      pip install --no-deps https://github.com/Theano/Theano/archive/rel-0.6.zip;
+      pip install "theano";
```
This can be installed with conda.

asmeurer replied:
It doesn't really matter (it only takes 4 seconds to pip install). I failed to notice that it has a conda build.

moorepants commented on sympy/core/function.py:
```diff
@@ -718,7 +718,7 @@ def _sage_(self):
         fname = str(self.func)
         args = [arg._sage_() for arg in self.args]
-        func = sage.function(fname, *args)
+        func = sage.function(fname)(*args)
```
Does this mean we no longer work with old versions of sage?

asmeurer replied:
No idea. It was giving a deprecation warning. I think we should only really bother with the latest version anyway, because the main idea is to work with Sage once it includes the latest version of SymPy inside of it.

isuruf commented:
About the last TODO regarding PyPy, you can use latest PyPy from PyPy's ppa with travis. See, https://github.com/mgedmin/mgp2pdf/commit/3f19b388421b61c5fb098bc436aae725aaced4b1

asmeurer commented:
@isuruf oh, I might try that, since it looks like it works with `sudo: false`.

asmeurer committed:
Try installing pypy from the PyPy ppa on Travis

Hopefully this fixes the errors, so I've optimistically removed the PyPy
builds from the allowed_failures.

isuruf commented:
@asmeurer, you have to create the virtualenv and activate with the new PyPy version as the new version is installed after the virtualenv is made with the older version.
Also `pypy-dev` package is not necessary since SymPy's not using the Python headers.

asmeurer committed:
Use conda to install theano in Travis

asmeurer committed:
Test autowrap in Python 3.5 on Travis

This adds yet another Travis build, but the autowrap builds finish in under a
minute so it shouldn't be too bad.

asmeurer committed:
Fix YAML parsing error in .travis.yml

I have no idea what was causing it, but this seems to fix it.

asmeurer committed:
pypy-dev shouldn't be needed in Travis

asmeurer committed:
Create a virtualenv for PyPy in Travis

asmeurer commented:
OK, all comments so far are addressed.

moorepants commented:
+1

asmeurer commented:
Unless there are more comments before then, I'll merge it when the tests come in (assuming I didn't break PyPy). Probably tomorrow.

asmeurer committed:
Fix indentation in .travis.yml

asmeurer committed:
Travis: Activate the PyPy virtualenv before installing SymPy

Otherwise it will install into the wrong virtualenv and it won't work.

asmeurer committed:
Properly skip numpy tests if numpy is not installed

asmeurer committed:
Try running the Travis autowrap tests in Python 3.4

Python 3.5 is apparently having some issues. The builds fail immediately with

$ source ~/virtualenv/python3.5_with_system_site_packages/bin/activate
/home/travis/build.sh: line 45: /home/travis/virtualenv/python3.5_with_system_site_packages/bin/activate: No such file or directory

asmeurer commented:
I'm guessing the failure at https://travis-ci.org/sympy/sympy/jobs/106575637 is an issue with Travis. I've tried working around it by using Python 3.4 for that build.

asmeurer committed:
Just add the Python 3.5 autowrap tests to the Travis allowed failures

I'm pretty sure this is failing because of some issues with Travis and I don't
really want to deal with this right now.

asmeurer committed:
Fix typo in .travis.yml

asmeurer committed:
Add the Sage tests to the Travis allow failures

Apparently the apt builds fail sometimes.

asmeurer commented:
The latest PyPy [still has bugs](https://travis-ci.org/sympy/sympy/jobs/106898871) apparently. So they're going back in the `allow_failures` list.

asmeurer committed:
Move PyPy builds to allow_failures on Travis

asmeurer commented:
W0000t!!!

moorepants commented:
:+1:
